ext/zend_test/test.c

```diff
@@ -720,7 +717,7 @@ static PHP_INI_MH(OnUpdateZendTestObserveOplineInZendMM)
 		ZT_G(zend_orig_heap) = zend_mm_get_heap();
 		zend_mm_set_heap(ZT_G(zend_test_heap));
 	} else if (ZT_G(zend_test_heap))  {
-		free(ZT_G(zend_test_heap));
+		zend_mm_shutdown(ZT_G(zend_test_heap), true, true);
```

Comment:
You need to set customer handlers to NULL before this call, and after that tests should pass

Reply:
```
=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
Bug #71527 Buffer over-write in finfo_open with malformed magic file [ext/fileinfo/tests/bug71527-mb.phpt]
Bug #71527 Buffer over-write in finfo_open with malformed magic file [ext/fileinfo/tests/bug71527.phpt]
Test proper watch comparisons when having multiple levels of indirection from a zval to its value [sapi/phpdbg/tests/watch_005.phpt]
Test multiple watch elements pointing to the same watchpoint [sapi/phpdbg/tests/watch_006.phpt]
=====================================================================
```

almost there :)

Reply:
Ah, the change from `malloc()` to `zend_mm_chunk_alloc_int()` when allocating the heap in `USE_ZEND_ALLOC=0` mode is breaking [this](https://github.com/php/php-src/blob/master/Zend/zend_alloc.c#L2457).

We could add a special case in __zend_free() and tracked_free(), such as:
```
if (ptr == AG(mm_heap)) {
    zend_mm_unmap(...);
} else {
    free(ptr);
}
```
but that's unfortunate.

Maybe we can change the shutdown code to this instead:

```
		if (shutdown) {
			shutdown(full, silent);
		} else if (full) {
			heap->custom_heap._free(heap ZEND_FILE_LINE_CC ZEND_FILE_LINE_EMPTY_CC);
		}
```

and then in https://github.com/php/php-src/blob/ab9a0295c566f9c34e395ffead8d030e60524cce/Zend/zend_alloc.c#L3070-L3071 we can add a shutdown callback to unmap the heap.

Since the shutdown hook was just introduced recently this should not be a breaking change. However this will break existing code relying on the `heap->custom_heap._free()` call to free the heap.

A BC-compatible alternative would be to use `aligned_alloc()` / `posix_memalign()` to allocate the heap, but this may not work on all supported systems.

@realFlowControl WDYT?

Reply:
It seems like your suggestion didn't solve the issue, and I'm a bit clueless about an elegant design to solve it :/

Reply:
I'm having a look right now, but as you mentioned ...
```C
if (ptr == AG(mm_heap)) {
    zend_mm_unmap(...);
} else {
    free(ptr);
}
```
... would not only be unfortunate, but also in the hot path.

> Maybe we can change the shutdown code to this instead:
> 
> ```C
> 		if (shutdown) {
> 			shutdown(full, silent);
> 		} else if (full) {
> 			heap->custom_heap._free(heap ZEND_FILE_LINE_CC ZEND_FILE_LINE_EMPTY_CC);
> 		}
> ```
> 
> and then in
> 
> https://github.com/php/php-src/blob/ab9a0295c566f9c34e395ffead8d030e60524cce/Zend/zend_alloc.c#L3070-L3071
> 
> we can add a shutdown callback to unmap the heap.

I am not sure I like the idea. The heap structure itself getting allocated via `malloc()` or `mmap()` is a implementation detail in the ZendMM and should not be of interest to any extension using `zend_mm_set_custom_handlers()`, additionally it is created if a custom handler is around or not. So far it was just coincidence, that we could just `free()` the heap as any other allocation. With this changed now, I would opt for the ZendMM cleaning this up on it's own and not relying on an extension to free internal structures.

I am currently playing with a few ideas, I'll report back

Reply:
So I'd suggest the following:

In `Zend/zend_alloc.c:zend_mm_shutdown()` we could instead of:
```C
		if (shutdown) {
			shutdown(full, silent);
		} else if (full) {
			ZEND_MM_RO_HEAP(heap)->custom_heap._free(heap ZEND_FILE_LINE_CC ZEND_FILE_LINE_EMPTY_CC);
		}
```
do this:
```C
		if (shutdown) {
			shutdown(full, silent);
		}
		if (full) {
			zend_mm_chunk_free(heap, heap->main_chunk, ZEND_MM_CHUNK_SIZE);
		}
```

This will ensure that the allocated heap structure will be freed (as it will be around anyway, no matter if a custom handler is around).

And we need to fix `ext/zend_test/test.c:PHP_RSHUTDOWN_FUNCTION(zend_test)` (Line 1332) from:
```C
		if (ZT_G(zend_test_heap))  {
			free(ZT_G(zend_test_heap));
			ZT_G(zend_test_heap) = NULL;
			zend_mm_set_heap(ZT_G(zend_orig_heap));
		}
```

to

```C
		if (ZT_G(zend_test_heap))  {
			zend_mm_set_custom_handlers(
				ZT_G(zend_test_heap),
				NULL,
				NULL,
				NULL
			);
			zend_mm_shutdown(ZT_G(zend_test_heap), true, true);
			ZT_G(zend_test_heap) = NULL;
			zend_mm_set_heap(ZT_G(zend_orig_heap));
		}
```

This later change is specifically to the problem that before this PR we where coincidently able to just `free()` a heap that was allocated from the ZendMM, but it is IMHO correct to call `zend_mm_shutdown()` on a heap that was allocated using `zend_mm_startup()` which is the case in this test extension.

WDYT? @arnaud-lb @jvoisin 

FYI: I tested locally with my proposed changes and tests seem stable

Reply:
Thank you @realFlowControl

 > I am not sure I like the idea. The heap structure itself getting allocated via `malloc()` or `mmap()` is a implementation detail in the ZendMM and should not be of interest to any extension using `zend_mm_set_custom_handlers()`, additionally it is created if a custom handler is around or not. So far it was just coincidence, that we could just `free()` the heap as any other allocation. With this changed now, I would opt for the ZendMM cleaning this up on it's own and not relying on an extension to free internal structures.

Yes, I agree

> So I'd suggest the following:
> 
> In `Zend/zend_alloc.c:zend_mm_shutdown()` we could instead of:
> 
> ```c
> 		if (shutdown) {
> 			shutdown(full, silent);
> 		} else if (full) {
> 			ZEND_MM_RO_HEAP(heap)->custom_heap._free(heap ZEND_FILE_LINE_CC ZEND_FILE_LINE_EMPTY_CC);
> 		}
> ```
> 
> do this:
> 
> ```c
> 		if (shutdown) {
> 			shutdown(full, silent);
> 		}
> 		if (full) {
> 			zend_mm_chunk_free(heap, heap->main_chunk, ZEND_MM_CHUNK_SIZE);
> 		}
> ```

:+1:

Reply:
Looking more into it, we should use `zend_mm_munmap(heap, REAL_PAGE_SIZE * 2)`, not `zend_mm_chunk_free()` here, for two reasons:
- We can to unmap `heap` itself
- `zend_mm_chunk_free()` may delegate to `heap->storage`

Reply:
Thanks @arnaud-lb, `zend_mm_munmap` is the way to go!